hxmkbib.1
.de d \" begin display
.sp
.in +4
.nf
..
.de e \" end display
.in -4
.fi
.sp
..
.TH "HXMKBIB" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"
.SH NAME
hxmkbib \- create bibliography from a template
.SH SYNOPSIS
.B hxmkbib
.RB "[\| " \-s
.IR separator " \|]"
.RB "[\| " \-a
.IR auxfile " \|]"
.RB "[\| " \-n
.IR maxauthors " \|]"
.RB "[\| " \-r
.IR moreauthors " \|]"
.IR bibfile " [\| " templatefile " \|]"
.SH DESCRIPTION
.LP
The
.B hxmkbib
commands reads a list of bibliographic keys (labels) from
.IR auxfile ,
finds the corresponding entries in
.I bibfile
and creates a bibliography, using
.I templatefile
as a model. The
.I auxfile
may, e.g., have been created by
.BR hxcite (1).
It consists of labels, one per line. The
.I bibfile
is a
.BR refer (1)
style database.
.B hxmkbib
looks for entries with a
.B %L
field equal to a key in the
.IR auxfile .
.PP
The
.I templatefile
consists of three parts:
.TP 10
.B preamble
The preamble is the part up to the first occurrence of
.BR %{ .
The preamble is copied to the output unchanged, except for occurrences 
of
.BR % .
To create a single % in the output, there must be two in the preamble
(%%). All other occurrences of % followed by another letter are not
copied, but are collected into a string called the "sort order." and
use to sort the entries, as explained below.
.TP
.B template
The template starts with
.B %{L:
and ends with a matching
.BR %} .
The text in between is copied as often as there are bibliographic
entries in
.I bibfile
that correspond to keys in 
.IR auxfile .
Variables in the template are replaced by the corresponding field in
the bibliographic entry: all occurrences of
.BI % x
will be replaced by the field
.BI % x
of the entry. Parts of the text may be enclosed in
.BI %{ x :
and
.BR %} .
This means that the text in between should only be output if the
current entry has a field
.IR x .
Text that is enclosed in
.BI %{! x :
and
.B %}
will only be output if the entry does
.B not
have a field
.IR x .
Both kinds of conditional sections may also be nested.
.TP
.B postamble
The text after the
.B %}
is copied unchanged to the output, after all bibliographic entries
have been processed.
.PP
By default bibliographic entries are copied to the output in the order 
of the keys in
.IR auxfile ,
except that keys that occur more than once are only used once. If the
preamble contains occurrences of
.BI % x
(where
.I x
is neither "%" nor "{") then these together determine the sort order.
E.g., if the preamble contains %A%D then the entries will be sorted
first on field A (author) and then on field D (date).
.PP
Here is an example of template file that creates a bibliography in
HTML format:
.d
<html>
<title>Bibliography</title>
<!--%A%D sorted on author, then date -->
<dl>
%{L:
<dt id="%L">%{A:%A%}%{!A:%{E:%E%}%{!E:%{Q:%Q%}%{!Q:-%}%}%}</dt>
<dd>%{B:"%T"
  in: %{E:%E (eds)
  %}<cite>%B.</cite>%{V: %V.%}
  %}%{J:"%T"
  in: %{E:%E (eds)
  %}<cite>%J.</cite>%{V: %V.%}%{N: %N.%}%{P: pp. %P.%}
  %}%{!B:%{!J:<cite>%T.</cite>
  %}%}%{I:%I.
  %}%{D:%D.
  %}%{C:%C.
  %}%{R:%R.
  %}%{S:%S.
  %}%{O:%O
  %}%{U:<a href="%U">%U</a>
  %}</dd>
%}
</dl>
</html>
.e
This template starts with four lines of preamble, including the sort
string %A%D on line 3. The sort string itself will not be output, but
the rest of the comment will.
.PP
From the line
.B %{L:
to the line
.B %}
is the template. E.g., the line that
starts with 
.B <dt id=...
contains a complex conditional text that prints the authors (%A) if
there are any, otherwise the editors (%E) if there are any, otherwise
the institution that is the author (%Q), if any, and a dash otherwise.
Note how the parts are nested, Most of the text is inside
.BR %{!A:...%} ,
meaning that that part will only be effective if there is no author
field (%A).
.PP
The final two lines are the postamble and will simply be copied
unchanged.
.PP
A bibliographic entry that looks like this in
.IR bibfile :
.d
%L Java
%A Gosling, James
%A Joy, Bill
%A Steele, Guy
%T The Java language specification
%D 1998
%I Addison-Wesley
%U http://java.sun.com/docs/books/jls/index.html
.e
will be printed by the template above as:
.d
<dt id="Java">Gosling, James; Joy, Bill; Steele, Guy</dt>
<dd><cite>The Java language specification.</cite>
  Addison-Wesley.
  1998.
  <a href="http://java.sun.com/docs/books/jls/index.html">http://java.sun.com/docs/books/jls/index.html</a>
  </dd>
.e
.SH OPTIONS
The following options are supported:
.TP 10
.BI \-a " auxfile"
The file that contains the list of keys (labels) for which
bibliographic entries should be printed. If the option is absent, the
name of this file is formed from the
.I templatefile
argument by removing the last extension and adding
.BR .aux .
If no
.I templatefile
is given, the default
.I auxfile
is
.BR aux.aux .
.TP
.BI \-s " separator"
If there are multiple authors or editors in an entry, their names will
be listed with a separator in between. By default the separator is ";
" (i.e., a semicolon and a space). With this option the separator can
be changed.
.TP
.BI \-n " maxauthors"
If there are more than
.I maxauthors
authors in an entry, only the first author will be printed and the
others will be replaced by the string
.IR moreauthors .
The default is 3.
.TP
.BI \-r " moreauthors"
The string to print if there are more than
.I maxauthors
authors. The default is "et al.".
.SH OPERANDS
The following operands are supported:
.TP 10
.I bibfile
The name of a bibliographic database must be given. It must be a file
in
.BR refer (1)
format and every entry must have at least a
.B %L
field, which is used as key. (Entries without such a field will be
ignored.)
.TP
.I templatefile
The name of the input file is optional. If absent,
.B hxmkbib
will read the template from stdin.
.SH "DIAGNOSTICS"
The following exit values are returned:
.TP 10
.B 0
Successful completion.
.TP
.B > 0
An error occurred. Usually this is because a file could not be opened
or because the %{ and %} pairs are not properly nested.
Very rarely it may also be an out of memory error. Some of the
possible error messages:
.TP
.I missing ':' in pattern
.B hxmkbib
found a %{ but the second or third letter after it was not a colon.
.TP
.I no '%{' in template file
The template file is unusable, because it contains no template.
.TP
.I unbalanced %{..%} in pattern
There are more %{ than %}.
.SH "SEE ALSO"
.BR asc2xml (1),
.BR hxcite (1),
.BR hxnormalize (1),
.BR hxnum (1),
.BR hxprune (1),
.BR hxtoc (1),
.BR hxunent (1),
.BR xml2asc (1),
.BR UTF-8 " (RFC 2279)"
.SH BUGS
Sorting is primitive: the program doesn't parse dates or names and
simply sorts "Jan 2000" under the letter "J" and "Albert Camus" under
the letter "A". For the moment the only work-around is to put names in
the
.I bibfile
as "Camus, Albert".
.PP
The program simply lists all authors or editors. There is no way to
generate an "et. al." after the third one. The work-around is to put
the "et. al." in the
.IR bibfile .
Putting commas between the first authors and the word "and" before the
final one is also not possible.
.PP
The program doesn't try to interpret names of authors or editors and
they cannot be reformatted. It is impossible to write a name that is
specified as "Sartre, Jean-Paul" in the
.I bibfile
as "J. Sartre" or as "Jean-Paul Sartre" in the output.
.PP
There is no way to suppress a period after a field if the field
already ends with a period. E.g., the template "%{A:A.%}" may generate
"A. Person Jr.." if the author is "A. Person Jr." The only option is
to either not put periods in the
.IR bibfile
or not put periods in the template.
.PP
Entries in the
.I bibfile
can only be used if they have a
.B %L
(label) field. The program cannot find entries by searching for
keywords, like
.BR refer (1).
.PP
.B hxmkbib
will replace any ampersands (&) and less-than (<) and greater-than (>)
signs that occur in the
.I bibfile
by their XML entities &amp; &lt; &gt; on the assumption that the
template is HTML/XML. This may not be appropriate for other formats.



ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxmkbib.c
/*
 * mkbib - extract database entries from a db and format them
 *
 * mkbib reads a refer-style database of bibliographic entries, a list
 * of keys and a pattern file and outputs a list of citations
 * formatted according to the pattern and optionally sorted.
 *
 * The keys must correspond to %L fields in the refer database.
 *
 * The pattern file has the following structure:
 *
 * pattern: PREAMBLE entry POSTAMBLE;
 * entry: "{L:" [ TEXT | FIELD | conditional ]* "}";
 * conditional: "{" !"? F ":" [ TEXT | FIELD | conditional ]* "}";
 *
 * In the output, the entry will be repeated as often as there are
 * unique keys. A FIELD is of the form "%x" and wil be replaced by
 * field x of the entry.
 *
 * A part of the form "{x:ZZZ}" will be replaced by ZZZ if field x
 * exists and by nothing otherwise. A part of the form "{!x:ZZZ}" will
 * be replaced by ZZZ if field x does not exist.
 *
 * Occurrences of %x in the preamble (where x is a field name) will
 * not be output, but serve to build up the sort order. The default
 * sort order is to keep entries in the order they occur in the
 * auxfile, but if, e.g., "%A%D%T" occurs in the preamble, entries
 * will be sorted on author, date and title.
 *
 * To insert a literal "{", "}" or "%" in the preamble or in an entry,
 * prefix them with "%": "%{", "%}" and "%%".
 *
 * Usage: mkbib [-a auxfile] bibfile [inputfile]
 *
 * bibfile is a refer-style database.
 *
 * inputfile is the file that serves as template. If absent, stdin
 * is read.
 *
 * -a auxfile gives the name of the list of keys. If absent, the name
 * will be the same as inputfile with the extension (if any)
 * changed to ".aux". If no inputfile is given the default auxfile
 * is "aux.aux". Duplicate keys will only be used once.
 *
 * Note: When the "{x:" and "}" are inside an HTML file, they may be
 * in places where data is not allowed. To make the input file
 * itself valid HTML, it may be necessary to put them inside comments:
 * <!--{x:--> and <!--}-->. If one of them is put inside a comment,
 * the other must be as well.
 *
 * Here is an example of an input file:
 *
 * <html>
 * <title>Bibliography</title>
 * <!-- sort order is Author, Date, Title %A%D%T-->
 * <dl>
 * <!--{L:--><dt id="%L">%L
 * <dd>{A:%A.} <em>{T:%T.}</em> {D:%D. }
 * <!--}--></dl>
 * </html>
 *
 * To do: if the template adds something like "(eds)", allow it to be
 * changed to "(ed)" if there is only one editor.
 *
 * Copyright © 1994-2004 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/copyright-software
 *
 * Author: Bert Bos <bert@w3.org>
 * Created: 19 March 2000
 * Version: $Id: hxmkbib.c,v 1.8 2018/02/15 19:04:22 bbos Exp $
 **/
#include "config.h"
#ifdef HAVE_ERRNO_H
#  include <errno.h>
#endif
#ifdef HAVE_UNISTD_H
#  include <unistd.h>
#endif
#include <stdio.h>
#if STDC_HEADERS
# include <string.h>
#else
# ifndef HAVE_STRCHR
#  define strchr index
#  define strrchr rindex
# endif
#endif
#include <stdlib.h>
#include <assert.h>

#ifdef HAVE_SEARCH_H
#  include <search.h>
#else
#  include "hash.e"		/* Use our own implementation */
#endif

#include <ctype.h>
#include <stdbool.h>
#include "heap.e"
#include "types.e"
#include "errexit.e"


#define LINESIZE 32768
#define INCR 25
/* Warning: arbitrary limit! */
#define HASHSIZE 4096				/* Size of hash table */

static string prog;				/* argv[0] */
static string sortorder = NULL;			/* Default is unsorted */
static string separator = "; ";			/* Separates authors */
static int et_al_limit = 3;			/* Max # of authors to print */
static string et_al = "et al.";			/* String if more authors */


/* escape -- print a string, escaping characters dangerous for XML/HTML */
static void escape(const string s, unsigned char *last)
{
  int i;

  for (i = 0; s[i]; i++)
    switch (s[i]) {
      case '<': printf("&lt;"); break;
      case '>': printf("&gt;"); break;
      case '&': printf("&amp;"); break;
      case '"': printf("&quot;"); break;
      default: putchar(s[i]);
    }
  if (i > 0) *last = s[i-1];
}


/* put_field -- copy field field of entry with label key */
static void put_field(const string key, unsigned char field, unsigned char *last)
{
  ENTRY *e, e1 = {key, NULL};
  string *lines;
  int i, j, nrfields;

  /* ToDo: escape dangerous characters */
  /* ToDo: print "et. al." if more than N authors */
  /* ToDo: for fields other than %A and %E use only the last occurrence */
  /* ToDo: interpret and pretty-print dates in a consistent manner */

  if (field == '%' || field == '{' || field == '}') { /* Literal */
    putchar(field);
    *last = '\0';
    return;
  }

  /* Find the entry for key */
  if (! (e = hsearch(e1, FIND))) {
    fprintf(stderr, "%s: entry for key %s not found\n", prog, key);
    return;
  }

  /* Count how many occurences of %field there are in the entry */
  lines = (string*)e->data;			/* Type cast */
  for (i = 0, nrfields = 0; lines[i]; i++)
    if (lines[i][1] == field) nrfields++;

  /* Check that there is indeed a field */
  if (nrfields == 0) {
    fprintf(stderr, "%s: entry %s has no field %%%c\n", prog, key, field);
    return;
  }

  /* Check that there are no duplicate fields, other than for A and E */
  if (nrfields != 1 && ! (field == 'A' || field == 'E')) {
    fprintf(stderr, "%s: entry %s has duplicate field %%%c\n",
	    prog, key, field);
    return;
  }

  /* Now print the field(s) */
  if (nrfields > et_al_limit) {			/* Print only the first */
    for (i = 0; lines[i][1] != field; i++);	/* Find the first */
    escape(lines[i] + 3, last);			/* Print with entities */
    printf("%s%s", separator, et_al);
    *last = et_al[strlen(et_al) - 1];
  } else {					/* Print all fields */
    for (i = 0, j = 0; lines[i]; i++) {
      if (lines[i][1] == field) {		/* Found it */
	if (j != 0) printf("%s", separator); /* Multiple fields */
	escape(lines[i] + 3, last);		/* Print with entities */
	j++;
      }
    }
  }
}


/* get_field -- check that entry for key has a field f, return ptr to field */
static string get_field(const string key, const unsigned char f)
{
  ENTRY *e, e1 = {key, NULL};
  string *lines;
  int i;

  /* Find the entry for key */
  e = hsearch(e1, FIND);
  assert(e != NULL);
  assert(e->data != NULL);

  /* Find a line that starts with %field */
  lines = (string*)e->data;			/* Type cast */
  for (i = 0; lines[i] && lines[i][1] != f; i++) ;

  assert(! lines[i] || (lines[i][0] == '%' && lines[i][2] == ' '));
  return lines[i];
}


/* compare_keys -- return the relative sort order for two keys: -1, 0, 1 */
static int compare_keys(const void *aptr, const void *bptr)
{
  ENTRY e, *ae, *be;
  int c, i;
  string af, bf, a = *(string*)aptr, b = *(string*)bptr;

  /* Get the entry for key a */
  e.key = a;
  ae = hsearch(e, FIND);
  assert(ae != NULL);

  /* Get the entry for key b */
  e.key = b;
  be = hsearch(e, FIND);
  assert(be != NULL);

  /* Loop over sortorder, stop as soon as entries a and b are unequal */
  for (i = 0, c = 0; c == 0 && sortorder[i]; i++) {
    af = get_field(a, sortorder[i]);
    bf = get_field(b, sortorder[i]);
    c = strcmp(af ? af : (string)"", bf ? bf : (string)"");
  }

  return c;
}


/* sort_keys -- sort the keys according to the sort order given */
static void sort_keys(string *keys, const int n)
{
  assert(sortorder != NULL);
  qsort(keys, n, sizeof(*keys), compare_keys);
}


/* conditional -- conditionally copy a %{...%} segment */
static int conditional(const string pattern, const string key,
		       unsigned char *last)
{
  bool on;
  int level, i = 1;

  /* Pattern starts with '{' */
  assert(pattern[0] == '{' && pattern[1] != '\0');

  /* Check the condition */
  if (pattern[i] == '!') on = !get_field(key, pattern[++i]);
  else on = get_field(key, pattern[i]) != NULL;

  if (pattern[i+1] != ':') errexit("%s: missing ':' in pattern\n", prog);

  /* Skip or copy until matching '%}' */
  if (! on) {					/* Skip until matching '}' */
    for (i += 2, level = 1; level != 0; i++)
      if (pattern[i] == '%') {
	if (pattern[++i] == '{') level++;
	else if (pattern[i] == '}') level--;
      }
    i--;					/* i points to '}' */
  } else {					/* Recursively copy segment */
    for (i += 2; true; i++)
      if (pattern[i] == '%') {
	if (pattern[++i] == '{') i += conditional(pattern + i, key, last);
	else if (pattern[i] == '}') break;
	else if (pattern[i] == '%') {putchar('%'); *last = '\0';}
	else put_field(key, pattern[i], last);
      } else if (*last != '.' || pattern[i] != '.') {
	putchar(pattern[i]);
	*last = '\0';
      } else {
	*last = '\0';				/* Don't print this '.' */
      }
  }
  
  return i;					/* Points at '}' */
}


/* copy -- copy pattern, expanding fields. (May sort keys) */
static void copy(const string pattern, string *keys, const int n)
{
  int j, start, end, level, slen = 0;
  unsigned char last = '\0';			/* Last char of field */

  assert(sortorder == NULL);

  /* ToDo: Find a way to declare the separator in the source. Maybe {&:...} */

  /* Find first '%{'. Also look for sort order */
  for (start = 0; pattern[start]; start++) {
    if (pattern[start] == '%') {		/* Special character */
      if (pattern[++start] == '{') {		/* Start of template */
	break;
      } else if ('A' <= pattern[start] && pattern[start] <= 'Z') {
	renewarray(sortorder, slen + 2);	/* Sort order */
	sortorder[slen] = pattern[start];
	sortorder[++slen] = '\0';
      } else {
	putchar('%');				/* Not special */
	putchar(pattern[start]);
      }
    } else {					/* Normal character */
      putchar(pattern[start]);
    }
  }
  if (!pattern[start]) {
    fprintf(stderr, "%s: warning: no '%%{' in input file\n", prog);
    return;					/* Nothing more to copy */
  }

  /* Sort the keys if there was a sort order */
  if (sortorder) sort_keys(keys, n);

  /* Start now points to '{'. Find matching '%}' */
  for (end = start + 1, level = 1; pattern[end] && level != 0; end++) {
    if (pattern[end] == '%') {
      if (pattern[++end] == '}') level--;
      else if (pattern[end] == '{') level++;
    }
  }
  if (level != 0) errexit("%s: unbalanced %{..%} in pattern\n", prog);

  /* End now points just after '}'. Loop over keys */
  for (j = 0; j < n; j++)
    conditional(pattern + start, keys[j], &last);
    
  /* Copy postamble */
  printf("%s", pattern + end);
}


/* in_list -- check if s is in the list of strings */
static bool in_list(const string s, const string *list, const int n)
{
  int i;

  for (i = 0; i < n && strcmp(s, list[i]) != 0; i++) ;
  return i < n;
}


/* read_keys -- read the list of keys from file f */
static string *read_keys(FILE *f, int *number)
{
  int i, e, n = 0;
  char line[LINESIZE];
  string *keys = NULL;

  clearerr(f);
  while (fgets(line, sizeof(line), f)) {

    /* Remove trailing \n and other whitespace */
    for (i = strlen(line); i > 0 && isspace(line[i-1]); i--) ;
    line[i] = '\0';

    /* ToDo: linear search fast enough? Books don't have 1000's of refs... */
    if (! in_list(line, keys, n)) {
      renewarray(keys, INCR * ((n + 1)/INCR + 1));
      keys[n++] = newstring(line);
    }
  }
  if ((e = ferror(f))) errexit("%s: %s\n", prog, strerror(e));

  *number = n;
  return keys;
}


/* check_and_store_entry -- check if we need this entry and if so store it */
static void check_and_store_entry(const string key, string *lines, int n)
{
  ENTRY e, *e1;

  renewarray(lines, INCR * ((n + 1)/INCR + 1));
  lines[n] = NULL;				/* Mark end of entry */
  if (key) {					/* Does it have a key at all */
    e.key = key;
    if ((e1 = hsearch(e, FIND))) 		/* Do we need this entry? */
      e1->data = (char*)lines;			/* Replace its data field */
  }
}


/* read_entries -- read the relevant entries from the refer database */
static void read_entries(FILE *f, const string *keys, const int n)
{
  char line[LINESIZE];
  string *lines = NULL;
  string key = NULL;
  ENTRY e, *e1;
  int i, j, fe;

  /* First enter all keys into the hash table without any data */
  for (i = 0; i < n; i++) {
    e.key = newstring(keys[i]);
    e.data = NULL;
    if (! hsearch(e, ENTER)) errexit("%s: %s\n", prog, strerror(errno));
  }

  /* Now read entries from the database */
  clearerr(f);
  i = 0;
  while (fgets(line, sizeof(line), f)) {

    if (line[0] != '%') {			/* Separator line */
      if (i != 0) {				/* We were in an entry */
	check_and_store_entry(key, lines, i);
	i = 0;					/* Reset */
	key = NULL;				/* Reset */
	lines = NULL;				/* Reset */
      }
    } else {					/* This line is a field */
      for (j = strlen(line); j > 0 && isspace(line[j-1]); j--) ;
      line[j] = '\0';				/* Remove trailing spaces */
      renewarray(lines, INCR * ((i + 1)/INCR + 1));
      lines[i] = newstring(line);
      if (strncmp(lines[i], "%L ", 3) == 0) key = lines[i] + 3;
      i++;
    }
  }
  if ((fe = ferror(f))) errexit("%s: %s\n", prog, strerror(fe));

  /* Check if last entry was already stored */
  if (i != 0)					/* We were still in an entry */
    check_and_store_entry(key, lines, i);

  /* Check that we found all keys */
  for (i = 0; i < n; i++) {
    e.key = keys[i];
    e1 = hsearch(e, FIND);
    assert(e1);
    if (! e1->data) errexit("%s: entry for \"%s\" not found\n", prog, keys[i]);
  }

}


/* read_pattern -- read the input file into memory */
static string read_pattern(FILE *f)
{
  string p = NULL;
  int n, len = 0;

  /* ToDo: use ferror to check for errors */
  do {
    renewarray(p, len + LINESIZE + 1);
    n = fread(p + len, sizeof(*p), LINESIZE, f);
    len += n;
  } while (! feof(f));
  p[len] = '\0';
  return p;
}


/* usage -- print usage message and exit */
static void usage(void)
{
  errexit("Version %s\nUsage: %s [-a auxfile] [-s sep] [-n maxauthors] [-r moreauthors] bibfile [inputfile]\n",
	  VERSION, prog);
}


/* main - main body */
int main(int argc, char *argv[])
{
  string auxfile = NULL, pattern, inputfile = NULL, dbfile, h;
  string *keys = NULL;
  FILE *f, *db, *aux;
  int c, n;

  /* Parse command line */
  prog = argv[0];
  while ((c = getopt(argc, argv, "a:s:n:r:")) != -1) {
    switch (c) {
    case 'a': auxfile = optarg; break;
    case 's': separator = optarg; break;
    case 'n': et_al_limit = atoi(optarg); break;
    case 'r': et_al = optarg; break;
    default: usage();
    }
  }
  if (optind == argc || argc > optind + 2) usage();

  /* First argument is refer database */
  dbfile = argv[optind++];

  /* Optional second argument is input file */
  if (optind != argc) inputfile = argv[optind];

  /* If we don't have an explicit auxfile yet, derive its name */
  if (! auxfile) {
    if (! inputfile) {
      auxfile = "aux.aux";
    } else {
      newarray(auxfile, strlen(argv[optind]) + 5);
      strcpy(auxfile, argv[optind]);
      if ((h = strrchr(auxfile, '.'))) *h = '\0';
      strcat(auxfile, ".aux");
    }
  }

  /* Create a hash table */
  if (! hcreate(HASHSIZE)) errexit("%s: not enough memory for hash table\n", prog);

  /* Read the pattern *before* the auxfile, so that we have a chance
     that the pipeline
     "hxcite -a aux bibfile file | hxmkib -a aux bibfile"
     works. I.e., that hxcite finished writing the auxfile before we
     start reading it. */

  /* Read pattern into memory */
  if (! (f = inputfile ? fopen(inputfile, "r") : stdin))
    errexit("%s: %s: %s\n", prog, inputfile, strerror(errno));
  pattern = read_pattern(f);
  if (fclose(f) != 0)
    errexit("%s: %s: %s\n", prog, inputfile, strerror(errno));

  /* Read keys from aux file */
  if (! (aux = fopen(auxfile, "r")))
    errexit("%s: %s: %s\n", prog, auxfile, strerror(errno));
  keys = read_keys(aux, &n);
  if (fclose(aux) != 0)
    errexit("%s: %s: %s\n", prog, auxfile, strerror(errno));

  /* Read the entries we need from the database */
  if (! (db = fopen(dbfile, "r")))
    errexit("%s: %s: %s\n", prog, dbfile, strerror(errno));
  read_entries(db, keys, n);
  if (fclose(db) != 0)
    errexit("%s: %s: %s\n", prog, dbfile, strerror(errno));

  /* Copy and expand the pattern */
  copy(pattern, keys, n);

  return 0;
}
